### DISCUSSION TITLE
Pipeline says package was published, but source is not updated

### DISCUSSION THREAD

**Discussion by fnando** [microsoft/winget-pkgs#192959] [General] [answered]
Hi! According to https://github.com/microsoft/winget-pkgs/blob/master/doc/FAQ.md#how-long-do-packages-take-to-be-published, packages are published every hour or so. 

[A package I submitted](https://github.com/microsoft/winget-pkgs/pull/192495) last Tuesday has a "Published Pipeline: Healthy" label, but I can't seem to find the latest version.

```console
PS C:\Users\fnando\Projects> date

Thursday, November 21, 2024 1:38:51 PM

PS C:\Users\fnando\Projects> winget source update && winget search --id Stellar.StellarCLI
Updating all sources...
Updating source: msstore...
Done
Updating source: winget...
Done
Name        Id                 Version     Source
--------------------------------------------------
Stellar CLI Stellar.StellarCLI 22.0.0-rc.1 winget

PS C:\Users\fnando\Projects> winget upgrade --id Stellar.StellarCLI
No available upgrade found.
No newer package versions are available from the configured sources.
```

Is the package really available or is that some aggressive caching on Windows side? Is there a way I can forcefully update the cache by removing some files?

Cheers!

**Comment by mdanish-kh** [comment, marked as answer]
The problem isn't with publishing, the version was published successfully which you can check with

```
 ~ winget show Stellar.StellarCLI --versions
Found Stellar CLI [Stellar.StellarCLI]
Version
-----------
22.0.0-rc.1
22.0.0
```

Why you're not seeing an available upgrade is because WinGet version sorting logic ranks `22.0.0-rc.1` higher than `22.0.0`. Until WinGet has better support for release channels ( tracked in issue https://github.com/microsoft/winget-cli/issues/147 ), I would suggest not publishing release candidate builds.

**Reply by fnando** [reply]
oof. Thanks!

What would be the correct naming pattern for rcs?

I assume I can remove that rc version from the manifest, right?

**Reply by leighmcculloch** [reply]
How do versions get sorted? Are versions sorted lexicographically?

**Reply by mdanish-kh** [reply]
> How do versions get sorted? Are versions sorted lexicographically?

It splits by character `.` and parses each "part". More granular details in:

https://github.com/microsoft/winget-cli/blob/master/doc/specs/%23980%20-%20Apps%20and%20Features%20entries%20version%20mapping.md#version-parsing-and-comparison-logic-in-winget

I think what we're seeing is governed by the rule: 

> _`else if string parts not equal, return comparison of strings`_


> What would be the correct naming pattern for rcs? I assume I can remove that rc version from the manifest, right?

Not sure I follow what you exactly mean by this. You should be able to make a PR to remove the release candidate version, and once that gets merged and published, users would get the stable `22.0.0` version when they do `winget install Stellar.StellarCLI`. For folks that have already installed `22.0.0-rc.1`, I'm afraid WinGet won't offer them an upgrade through `winget upgrade` regardless

**Reply by leighmcculloch** [reply]
@mdanish-kh Thanks for the link. We found a similar discussion of the same details with examples here at this link: https://github.com/microsoft/winget-pkgs/blob/master/doc/Authoring.md#version-sorting-in-winget

It looks like from that document that `rc` versions are intended to sort after their non-rc targets.

The following example is provided:
https://github.com/microsoft/winget-pkgs/blob/2286e89b81ea5b7a87529ee692495db51b3b3be7/doc/Authoring.md?plain=1#L154-L157

It says that `1.2` takes precedence over `1.2-rc`.

I think the issue we have is that our RC version has a `.1` suffix. And winget is treating that final segment like a new version part, when it shouldn't be interpreted that way according to the semver specification: https://semver.org.

Winget interprets `22.0.0` to be later than `22.0.0-rc`. But Winget then interprets `22.0.0-rc.1` to be later than `22.0.0`. According to semver, `22.0.0` is still later than `22.0.0-rc.1`. Everything after the `-` should be interpreted and sorted independently of everything before the `-`.

I understand winget may not implement semver completely, but it could benefit by doing so since so many other ecosystems already use it.

**Reply by mdanish-kh** [reply]
Hmmm...interesting. I'll add @denelon & @Trenly  to the conversation if they can share their thoughts as they could be more knowledgeable on the client's version sorting areas

**Reply by Trenly** [reply]
I'll look into this more tomorrow, but my guess would also be that it is the extra `.1` causing the issues. Specifically, I know that if one version has more parts, it will be considered newer, which is probably what is happening here. It sees that all the integers are the same, but didn't catch that one of the preceding parts should have caused it to return early before even comparing the last part

**Reply by Trenly** [reply]
For my own reference when I get time to dive into where the issue is, here is the code that compares the two versions -
* https://github.com/microsoft/winget-cli/blob/25ccc2a6db72b88daab6447668f1437eb61649ad/src/AppInstallerSharedLib/Versions.cpp#L140-L171

In this case, it should be hitting the path that the parts are not equal, so that would be in the part comparator -
* https://github.com/microsoft/winget-cli/blob/25ccc2a6db72b88daab6447668f1437eb61649ad/src/AppInstallerSharedLib/Versions.cpp#L324-L352

**Reply by Trenly** [reply]
I did confirm that this seems to be a bug in the client by adding a new test, which caused it to fail -
```
E:\winget-cli\src\AppInstallerCLITests\Versions.cpp(122): FAILED:
  REQUIRE( vA < vB )
with expansion:
  {?} < {?}

===============================================================================
test cases:  1 |  0 passed | 1 failed
assertions: 71 | 70 passed | 1 failed
```

**Reply by Trenly** [reply]
Oh my. This is a weird bug. 

```
    RequireLessThan("1-rc", "1"); // Test case passes
    RequireLessThan("1.2-rc", "1.2"); // Test case passes
    RequireLessThan("1.0-rc", "1.0"); // Test case fails
```

Edit: I'm very confused now. When comparing the individual parts, the parts are compared just fine
```
    RequirePartLessThan("2-rc", "2"); // Test case passes
    RequirePartLessThan("0-rc", "0"); // Test case passes
```

**Reply by Trenly** [reply]
I'm just going to keep leaving myself notes here so that I don't forget. I just had a thought.

When a version is created, all trailing zeroes are trimmed off. This means that `22.0.0`, when broken into parts, is actually stored as `[22]` and not as `[22, 0, 0]`. This means that it's hitting the condition where `i >= other.m_parts.size()` and is just breaking the for loop instead of actually trying to pad with zeroes when comparing

Edit: Confirmed by removing the trimming, which caused the sorting to be correct. This isn’t a real fix though. . .

**Reply by Trenly** [reply]
PR has been raised - https://github.com/microsoft/winget-cli/pull/5001
